### PR TITLE
Fix SSH sessions recorded on proxy not being fully closed

### PR DESCRIPTION
**Pull request by gabrielcorado**
This issue affects only SSH sessions recorded at Proxy. When the `session.Close` is called from the Proxy, the session is not properly closed, leaving it open.

The problem was related to the session waiting for the parties to be closed (`s.term.AddParty(-1)`). This PR updates the close flow from `session.Close` and `remoteTerminal.Close` to fix it (commented on code).

### CODE REVIEW

File: lib/srv/sess.go [hunk 1]
```diff
@@ -910,18 +910,19 @@ func (s *session) haltTerminal() {
...
 	s.BroadcastMessage("Closing session...")
 	s.log.Infof("Closing session")
 
-	serverSessions.Dec()
-
-	// Remove session parties and close client connections.
+	// Remove session parties and close client connections. Since terminals
+	// might await for all the parties to be released, we must called it first.
+	// Closing the parties will cause their SSH channel to be closed, meaning
+	// any gorouting reading from it will be released.
```
Comment (zmb3) [inline lib/srv/sess.go:919]:
```suggestion
	// any goroutine reading from it will be released.
```

File: lib/srv/sess_test.go [hunk 2]
```diff
@@ -988,3 +993,176 @@ func TestSessionRecordingMode(t *testing.T) {
...
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+
+	go func() {
+		conn, err := listener.Accept()
+		if err != nil {
+			t.Logf("error while accepting ssh connections: %s", err)
+		}
+
+		srvConn, chCh, reqCh, err := ssh.NewServerConn(conn, cfg)
```
Comment (zmb3) [inline lib/srv/sess_test.go:1111]: Check the error?

File: lib/srv/sess_test.go [hunk 3]
```diff
@@ -988,3 +993,176 @@ func TestSessionRecordingMode(t *testing.T) {
...
+
+	// Establish a connection to the newly created server.
+	sessCh := make(chan *tracessh.Session)
+	go func() {
+		client, err := tracessh.Dial(ctx, listener.Addr().Network(), listener.Addr().String(), &ssh.ClientConfig{
+			User:            "user",
+			Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
+			HostKeyCallback: ssh.FixedHostKey(signer.PublicKey()),
+		})
+		require.NoError(t, err)
```
Comment (espadolini) [inline lib/srv/sess_test.go:1152]: `require` in this specific goroutine works but the outer goroutine is left waiting for a timeout. I don't see a sensible way to use `ssh.Dial` (like, ever), so maybe the solution is to split the tcp dialing (with a timeout, and a timer that kills the connection after a while) and the SSH protocol stuff, and just do everything in a single goroutine?

```go
timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
defer cancel()

nc, err := new(net.Dialer).DialContext(timeoutCtx, listener.Addr().Network(), listener.Addr().String())
require.NoError(t, err)
defer nc.Close()
stop := context.AfterFunc(timeoutCtx, func() { _ = nc.Close() })
defer stop()

huey, dewey, louie, err := ssh.NewClientConn(nc, listener.Addr().String(), &ssh.ClientConfig{...})
require.NoError(t, err)
client := ssh.NewClient(huey, dewey, louie)

sess, err := client.NewSession()
require.NoError(t, err)
require.NoError(t, sess.Close())
```

**Review by espadolini — APPROVED** [review]
LGTM other than the test code

File: lib/srv/sess_test.go [hunk 4]
```diff
@@ -988,3 +993,176 @@ func TestSessionRecordingMode(t *testing.T) {
...
+
+	reg, _ := NewSessionRegistry(SessionRegistryConfig{
+		Srv:                   srv,
+		SessionTrackerService: srv.auth,
+	})
+	t.Cleanup(func() { reg.Close() })
+
+	scx := newTestServerContext(t, reg.Srv, nil)
+	sid := string(rsession.NewID())
+	scx.SetEnv(sshutils.SessionEnvVar, sid)
```
Comment (greedy52) [inline lib/srv/sess_test.go:1009]: What does this `sid` do?

File: lib/srv/sess_test.go [hunk 5]
```diff
@@ -988,3 +993,176 @@ func TestSessionRecordingMode(t *testing.T) {
...
+
+	select {
+	case err := <-closeChan:
+		require.NoError(t, err)
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "expected session to be closed")
+	}
+}
+
+func mockSSHSession(t *testing.T) *tracessh.Session {
```
Comment (greedy52) [inline lib/srv/sess_test.go:1090]: `t.Helper()`

**Review by greedy52 — APPROVED** [review]

File: lib/srv/sess_test.go (hunk 5 above)
Comment (greedy52) [reply lib/srv/sess_test.go:1087]: you might have added `t.Helper()` in the wrong function =D

Comment (gabrielcorado) [reply lib/srv/sess_test.go:1085]: True, I've updated it. Thanks!

File: lib/srv/sess_test.go (hunk 3 above)
Comment (espadolini) [reply lib/srv/sess_test.go]: The timeout is not applied on the SSH handshake, which is the flaw that makes `ssh.Dial` basically unusable in real production code.

**Pull request merged**